src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt

```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
+      val outputBuffer = PrintStream(execution)
+      return WorkerIO(inputBuffer, outputBuffer, execution) {}
+    }
+  }
+
+  override fun close() {
+    restore.invoke()
   }
 }
 
-class InvocationWorker(
-  private val delegate: CommandLineProgram,
-  private val buffer: ByteArrayOutputStream
-) {
+/** PersistentWorker follows the Bazel worker protocol and executes a CommandLineProgram. */
+class PersistentWorker(
+  private val io: WorkerIO,
+  private val program: CommandLineProgram
+) : Worker {
+  val logger = Logger.getLogger(PersistentWorker::class.java.canonicalName)
+
+  enum class Status {
+    OK, INTERRUPTED, ERROR
+  }
+
+  override fun run(args: List<String>): Int {
+    while (true) {
+      val request = WorkRequest.parseDelimitedFrom(io.input) ?: continue
 
-  fun invoke(args: List<String>): Pair<Int, String> = WorkingDirectoryContext.newContext()
-      .use { wdCtx ->
-        return try {
-          delegate.apply(wdCtx.dir, args)
-        } catch (e: RuntimeException) {
-          if (e.wasInterrupted()) BazelWorker.INTERRUPTED_STATUS
-          else BazelWorker.ERROR_STATUS.also {
-            System.err.println(
-                "ERROR: Worker threw uncaught exception with args: ${args}"
-            )
-            e.printStackTrace(System.err)
+      val (status, exit) = WorkingDirectoryContext.newContext()
+        .runCatching {
+          request.argumentsList
+            ?.let {
+              maybeExpand(it)
+            }
+            .run {
+              Status.OK to program.apply(dir, maybeExpand(request.argumentsList))
+            }
+        }
+        .recover { e: Throwable ->
+          io.execution.write((e.message ?: e.toString()).toByteArray(UTF_8))
+          if (!e.wasInterrupted()) {
+            logger.log(Level.SEVERE,
+                       "ERROR: Worker threw uncaught exception",
+                       e)
+            Status.ERROR to 1
+          } else {
+            Status.INTERRUPTED to 1
           }
-        } to buffer.toString()
+        }
+        .getOrThrow()
+
+      val response = with(WorkResponse.newBuilder()) {
+        output = String(io.execution.toByteArray(), UTF_8)
+        exitCode = exit
+        setRequestId(request.requestId)
+      }.build()
+
+      // return the response
+      response.writeDelimitedTo(io.output)
+      io.output.flush()
+
+      // clear execution logs
+      io.execution.reset()
+
+      if (status == Status.INTERRUPTED) {
+        break
       }
+    }
+    logger.info("Shutting down worker.")
+    return 0
+  }
 }
+
+class InvocationWorker(
+  private val io: WorkerIO,
+  private val program: CommandLineProgram
+) : Worker {
+  val logger: Logger = Logger.getLogger(InvocationWorker::class.java.canonicalName)
+  override fun run(args: List<String>): Int {
+    return WorkingDirectoryContext.newContext().runCatching {
+      program.apply(dir, maybeExpand(args))
+    }.recover { e ->
+      logger.log(Level.SEVERE,
```

Comment:
Nit. static import Level and that can be a one-liner.

Makes it a wee bit easier to see what's happening with the return code. Also, is this a place where oen of those `return@`'s will work?

Reply:
Still too long for one line, I'm afraid.